## Contents
{:.no_toc}

-   [Additional materials](#additional-materials)
-   [Session information](#session-information)
{:toc} \# Additional materials and session information {#sec-si}

### Additional materials

-   The [Single-cell proteomics data analysis using `QFeatures` and
    `scp`](https://lgatto.github.io/QFeaturesScpWorkshop2021/) workshop
    is provided as two vignettes. The first one provides a general
    introduction to the `QFeatures` class in the general context of mass
    spectrometry-based proteomics data manipulation. The second vignette
    focuses on single-cell application and introduces the `scp` package
    \[@Vanderaa:2021\] as an extension of `QFeatures`. This second
    vignette also provides exercises that give the attendee the
    opportunity to apply the learned concepts to reproduce a published
    analysis on a subset of a real data set.

-   The [Seamless Integration of Mass Spectrometry Data from Different
    Sources](https://jorainer.github.io/SpectraTutorials/) workshop
    explains the `Spectra` package and shows how this new infrastructure
    can be used to represent and analyze Mass Spectrometry (MS) data.
    The use of different data type and format specific backends
    guarantees an easy expandability of the infrastructure. This is
    illustrated in this tutorial on a simple use case, in which MS/MS
    (a.k.a. MS2) spectra, measured by an MS instrument, are compared and
    matched against a public spectral database containing public domain
    reference MS2 spectra.

-   A tutorial presenting [Use Cases and Examples for Annotation of
    Untargeted Metabolomics
    Data](https://jorainer.github.io/MetaboAnnotationTutorials/) using
    the `MetaboAnnotation` and `MetaboCoreUtils` packages
    \[@Rainer:2022\].

-   A tutorial for [LC-MS pre-processing with
    `xcms`](https://jorainer.github.io/metabolomics2018/xcms-preprocessing.html)
    with examples on how to perform centroiding of profile-mode MS data
    in `R` and how to define data set-specific parameters for
    chromatographic peak detection, alignment and correspondence of
    untargeted (LC-MS-based) metabolomics data.

### Session information

The following packages have been used to generate this document.

    sessionInfo()

    ## R version 4.2.0 (2022-04-22)
    ## Platform: x86_64-pc-linux-gnu (64-bit)
    ## Running under: Ubuntu 20.04.4 LTS
    ## 
    ## Matrix products: default
    ## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
    ## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3
    ## 
    ## locale:
    ##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
    ##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
    ##  [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
    ## 
    ## attached base packages:
    ## [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     
    ## 
    ## other attached packages:
    ##  [1] mzID_1.34.0                 patchwork_1.1.1             factoextra_1.0.7           
    ##  [4] gplots_3.1.3                limma_3.52.2                forcats_0.5.1              
    ##  [7] stringr_1.4.0               purrr_0.3.4                 readr_2.1.2                
    ## [10] tidyr_1.2.0                 tibble_3.1.7                tidyverse_1.3.1            
    ## [13] ggplot2_3.3.6               dplyr_1.0.9                 PSMatch_1.0.0              
    ## [16] MsCoreUtils_1.8.0           QFeatures_1.6.0             MultiAssayExperiment_1.22.0
    ## [19] SummarizedExperiment_1.26.1 Biobase_2.56.0              GenomicRanges_1.48.0       
    ## [22] GenomeInfoDb_1.32.2         IRanges_2.30.0              MatrixGenerics_1.8.0       
    ## [25] matrixStats_0.62.0          Spectra_1.6.0               ProtGenerics_1.28.0        
    ## [28] BiocParallel_1.30.3         S4Vectors_0.34.0            BiocGenerics_0.42.0        
    ## [31] mzR_2.30.0                  Rcpp_1.0.8.3                BiocStyle_2.24.0           
    ## [34] rmarkdown_2.14             
    ## 
    ## loaded via a namespace (and not attached):
    ##   [1] readxl_1.4.0            backports_1.4.1         plyr_1.8.7              igraph_1.3.2           
    ##   [5] gmm_1.6-6               lazyeval_0.2.2          digest_0.6.29           msmbstyle_0.0.19       
    ##   [9] foreach_1.5.2           htmltools_0.5.2         fansi_1.0.3             magrittr_2.0.3         
    ##  [13] cluster_2.1.3           doParallel_1.0.17       tzdb_0.3.0              remotes_2.4.2          
    ##  [17] modelr_0.1.8            imputeLCMD_2.1          sandwich_3.0-2          prettyunits_1.1.1      
    ##  [21] colorspace_2.0-3        rvest_1.0.2             ggrepel_0.9.1           haven_2.5.0            
    ##  [25] xfun_0.31               callr_3.7.0             crayon_1.5.1            RCurl_1.98-1.7         
    ##  [29] jsonlite_1.8.0          impute_1.70.0           zoo_1.8-10              iterators_1.0.14       
    ##  [33] glue_1.6.2              gtable_0.3.0            zlibbioc_1.42.0         XVector_0.36.0         
    ##  [37] DelayedArray_0.22.0     car_3.1-0               pkgbuild_1.3.1          abind_1.4-5            
    ##  [41] scales_1.2.0            mvtnorm_1.1-3           DBI_1.1.3               rstatix_0.7.0          
    ##  [45] clue_0.3-61             httr_1.4.3              ellipsis_0.3.2          pkgconfig_2.0.3        
    ##  [49] XML_3.99-0.10           farver_2.1.0            dbplyr_2.2.0            utf8_1.2.2             
    ##  [53] tidyselect_1.1.2        labeling_0.4.2          rlang_1.0.2             munsell_0.5.0          
    ##  [57] cellranger_1.1.0        tools_4.2.0             cli_3.3.0               msdata_0.36.0          
    ##  [61] generics_0.1.2          broom_0.8.0             evaluate_0.15           fastmap_1.1.0          
    ##  [65] yaml_2.3.5              processx_3.6.1          knitr_1.39              fs_1.5.2               
    ##  [69] caTools_1.18.2          AnnotationFilter_1.20.0 ncdf4_1.19              xml2_1.3.3             
    ##  [73] compiler_4.2.0          rstudioapi_0.13         curl_4.3.2              ggsignif_0.6.3         
    ##  [77] reprex_2.0.1            stringi_1.7.6           highr_0.9               ps_1.7.1               
    ##  [81] lattice_0.20-45         Matrix_1.4-1            tmvtnorm_1.5            vctrs_0.4.1            
    ##  [85] pillar_1.7.0            norm_1.0-10.0           lifecycle_1.0.1         BiocManager_1.30.18    
    ##  [89] bitops_1.0-7            R6_2.5.1                pcaMethods_1.88.0       bookdown_0.27          
    ##  [93] KernSmooth_2.23-20      codetools_0.2-18        MASS_7.3-57             gtools_3.9.2.2         
    ##  [97] assertthat_0.2.1        rprojroot_2.0.3         withr_2.5.0             GenomeInfoDbData_1.2.8 
    ## [101] parallel_4.2.0          hms_1.1.1               grid_4.2.0              carData_3.0-5          
    ## [105] ggpubr_0.4.0            lubridate_1.8.0
